src/common/com/intellij/plugins/haxe/haxelib/HaxeLibrary.java

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright 2017 Eric Bishton
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intellij.plugins.haxe.haxelib;
+
+import com.intellij.openapi.projectRoots.Sdk;
+import com.intellij.openapi.roots.OrderRootType;
+import com.intellij.openapi.roots.libraries.Library;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.plugins.haxe.lang.psi.HaxeClass;
+import com.intellij.plugins.haxe.util.HaxeDebugLogger;
+import com.intellij.plugins.haxe.util.HaxeFileUtil;
+import org.apache.commons.lang.NotImplementedException;
+import org.apache.log4j.Level;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.security.InvalidParameterException;
+import java.util.*;
+
+public class HaxeLibrary {
+
+  private static HaxeDebugLogger LOG = HaxeDebugLogger.getLogger();
+  static {LOG.setLevel(Level.DEBUG);}  // Remove when finished debugging.
+
+  private static String CURRENT_DIR = ".";
+
+  final private HaxelibLibraryCache myCache;
+  final private String myName;
+  final private String myLibraryRoot;
+  final private String myRelativeClasspath;
+  final private HaxelibMetadata myMetadata;
+  final private HaxelibSemVer mySemVer;
+  // TODO: Add the extraParams.hxml data here.  Use the hxml parser; see LimeUtil.getLimeProjectModel() as an example.
+
+
+  private HaxeLibrary(@NotNull String name, @NotNull VirtualFile libraryRoot, @NotNull HaxelibLibraryCache owner) {
+    myCache = owner;
+    myLibraryRoot = libraryRoot.getUrl();
+
+    myMetadata = HaxelibMetadata.load(libraryRoot);
+    HaxeLibraryInfo pathInfo = HaxelibUtil.deriveLibraryInfoFromPath(owner.getSdk(), libraryRoot.getPath());
+
+    String mdname = myMetadata.getName();
```

Comment:
myMetadata could be null and this entry will throw NullPointerException (checked today)

Reply:
You are right, on the face of it.  I will add a check there.

In actual practice, I'm not sure that this can happen without the library being corrupted.  Haxelib requires a haxelib.json file in order to publish, distribute, or convert a library into a classpath (which it does for the compiler).  If there is no haxelib.json, then haxelib won't be able to load the library and it can't be used by the compiler.  Of course, that doesn't mean that we should crash, either.

Reply:
One can `$ haxelib dev mylib path/to/sources` without having `haxelib.json` there.